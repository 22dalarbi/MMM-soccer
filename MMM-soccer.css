.icon {
    width: 24px;
    height: 24px;
    margin-right: 8px;
    -webkit-filter: grayscale(100%);

}

.row {
    display: flex;
    align-items: center;
    margin: 4px 0px;
}

.keys {
    font-weight: bold;
    justify-content: flex-end;
}

.key-icon {
    height: 20px;
    width: 20px;
    -webkit-filter: invert(100%) grayscale(100%);
}

.name {
    width: 210px;
    text-align: left;
}

.details {
    display: flex;
    justify-content: space-between;
    width: 50px;
}

h4 {
    margin-bottom: 10px;
}