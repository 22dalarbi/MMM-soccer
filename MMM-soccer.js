/* Magic Mirror
 * Module: MMM-soccer
 *
 * By fewieden https://github.com/fewieden/MMM-soccer
 *
 * MIT Licensed.
 */

Module.register("MMM-soccer",{

    defaults: {
        api_key: false,
        show: 'GERMANY',
        focus_on: false,  // false or the name of a team to focus on (used with max_teams)
        max_teams: false,   // false or the number of teams to show either side of the focused team
        leagues: {
            "GERMANY":430,
            "FRANCE": 434,
            "ENGLAND": 426,
            "SPAIN": 436,
            "ITALY": 438 
        }
    },

    // A loading boolean.
    loading: true,

    // Subclass start method.
    start: function() {
        Log.info("Starting module: " + this.name);
        this.currentLeague = this.config.leagues[this.config.show];
        this.getData();

        this.getData();
        var self = this; 
        setInterval(function() {
            self.getData();
        }, this.config.api_key ? 300000 : 1800000); // with api_key every 5min without every 30min
    },

    getData: function(){
        this.sendSocketNotification('GET_DATA', {league: this.currentLeague, api_key: this.config.api_key});
    },

    // Subclass socketNotificationReceived method.
    socketNotificationReceived: function(notification, payload){
        if(notification === 'DATA'){
            this.standing = payload;
            this.loading = (!this.standing);
            this.updateDom();
        }
    },

    // Subclass getStyles method.
    getStyles: function() {
        return ["MMM-soccer.css"];
    },

    // Override dom generator.
    getDom: function() {
        var wrapper = document.createElement("div");

        if (this.loading ||
            !this.standing) {
            var title = document.createElement("header");
            title.innerHTML = this.name;
            wrapper.appendChild(title);

            var subtitle = document.createElement("div");
            subtitle.className = "small dimmed light";
            subtitle.innerHTML = (this.loading) ? this.translate("LOADING") : "No data available" ;
            wrapper.appendChild(subtitle);

            return wrapper;
        }

        // Generate Standings Table
        if(this.standing){
            // League header
            var title = document.createElement("header");
            title.innerHTML = this.standing.leagueCaption;
            wrapper.appendChild(title);

            // Matchday indicator
            var subtitle = document.createElement("p");
            subtitle.classList.add("xsmall");
            subtitle.innerHTML = "Matchday: " + this.standing.matchday;
            wrapper.appendChild(subtitle);

            // Standings container
            var rows = document.createElement('div');
            rows.classList.add("xsmall");

            // Standings header row
            var row = document.createElement('div');
            row.classList.add('row', 'keys');

            var name = document.createElement('div');
            name.classList.add('name');
            name.innerHTML = 'Team';
            row.appendChild(name);

            var details = document.createElement('div');
            details.classList.add('details');

            var points = document.createElement('img');
            points.classList.add('key-icon');
            points.src = this.file('icons/points.png');
            details.appendChild(points);

            var goals = document.createElement('img');
            goals.classList.add('key-icon');
            goals.src = this.file('icons/goal.png');
            details.appendChild(goals);
            row.appendChild(details);

            rows.appendChild(row);

            // Get First and Last teams to display in standings
            var focusTeamIndex;
            var firstTeam = 0;
            var lastTeam = this.standing.standing.length;
            if (this.config.max_teams >= 0 && this.config.focus_on) {
                for(var i = 0; i < this.standing.standing.length; i++){
                    if(this.standing.standing[i].teamName === this.config.focus_on){
                        focusTeamIndex = i;
                        firstTeam = (focusTeamIndex - this.config.max_teams < 0) ? firstTeam : focusTeamIndex - this.config.max_teams;
                        lastTeam = (focusTeamIndex + this.config.max_teams + 1 > lastTeam) ? lastTeam : focusTeamIndex + this.config.max_teams + 1;
                        break;
                    }

                }
            }

            // Render Team Rows
            for(var i = firstTeam; i < lastTeam; i++){
                var row = document.createElement('div');
                row.classList.add('row');

                var pos = document.createElement('div');
                pos.classList.add('position');
                pos.innerHTML = this.standing.standing[i].position;
                row.appendChild(pos);

                var icon = document.createElement('img');
                icon.classList.add('icon');
                if (this.standing.standing[i].crestURI != "null") 
                    icon.src = this.standing.standing[i].crestURI;   // API returns "null" for teams without a crest
                row.appendChild(icon);

                var name = document.createElement('div');
                name.classList.add('name');
                name.innerHTML = this.standing.standing[i].teamName;
                row.appendChild(name);

                var details = document.createElement('div');
                details.classList.add('details');

                var points = document.createElement('span');
                points.innerHTML = this.standing.standing[i].points;
                details.appendChild(points);

                var goals = document.createElement('span');
                goals.innerHTML = this.standing.standing[i].goalDifference;
                details.appendChild(goals);
                row.appendChild(details);

                // Create fade in/out effect.
                if (this.config.max_teams && focusTeamIndex >= 0) {
                    if (i != focusTeamIndex) {
                        var currentStep = Math.abs(i - focusTeamIndex);
                        row.style.opacity = 1 - (1 / this.config.max_teams * currentStep);
                    }
                }

                rows.appendChild(row);
            }
            wrapper.appendChild(rows);
        }
        return wrapper;
    }
});
