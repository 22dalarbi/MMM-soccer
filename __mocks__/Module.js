global.Module = {
    definitions: {},
    create(name) {
        return this.definitions[name];
    },
    register(name, overrides) {
        const base = {
            requiresVersion: '2.0.0',

            defaults: {},

            showHideTimer: null,

            lockStrings: [],

            _nunjucksEnvironment: {addFilter: jest.fn()},

            init() {
                Log.log(this.defaults);
            },

            start() {
                Log.info('Starting module: ' + this.name);
            },

            getScripts() {
                return [];
            },

            getStyles() {
                return [];
            },

            getTranslations() {
                return false;
            },

            getDom() {},

            getHeader() {
                return this.data.header;
            },

            getTemplate() {
                return '<div class="normal">' + this.name + '</div><div class="small dimmed">' + this.identifier + '</div>';
            },

            getTemplateData() {
                return {};
            },

            notificationReceived(notification, payload, sender) {
                if (sender) {
                    Log.log(this.name + ' received a module notification: ' + notification + ' from sender: ' + sender.name);
                } else {
                    Log.log(this.name + ' received a system notification: ' + notification);
                }
            },

            nunjucksEnvironment() {
                return this._nunjucksEnvironment;
            },

            socketNotificationReceived(notification, payload) {
                Log.log(this.name + ' received a socket notification: ' + notification + ' - Payload: ' + payload);
            },

            suspend() {
                Log.log(this.name + ' is suspended.');
            },

            resume() {
                Log.log(this.name + ' is resumed.');
            },

            setData(data) {
                this.data = data;
                this.name = data.name;
                this.identifier = data.identifier;
                this.hidden = false;

                this.setConfig(data.config, data.configDeepMerge);
            },

            setConfig(config) {
                this.config = Object.assign({}, this.defaults, config);
            },

            socket() {},

            file(file) {
                return (this.data.path + '/' + file).replace('//', '/');
            },

            loadStyles(callback) {
                this.loadDependencies('getStyles', callback);
            },

            loadScripts(callback) {
                this.loadDependencies('getScripts', callback);
            },

            loadDependencies(funcName, callback) {},

            loadTranslations(callback) {},

            translate(key, defaultValueOrVariables, defaultValue) {},

            updateDom(speed) {},

            sendNotification(notification, payload) {},

            sendSocketNotification: jest.fn(),

            hide(speed, callback, options) {
                if (typeof callback === 'object') {
                    options = callback;
                    callback = function () {
                    };
                }

                callback = callback || function () {
                };
                options = options || {};

                const self = this;
                MM.hideModule(
                    self,
                    speed,
                    () => {
                        self.suspend();
                        callback();
                    },
                    options
                );
            },

            show(speed, callback, options) {
                if (typeof callback === 'object') {
                    options = callback;
                    callback = function () {
                    };
                }

                callback = callback || function () {
                };
                options = options || {};

                const self = this;
                MM.showModule(
                    this,
                    speed,
                    () => {
                        self.resume();
                        callback();
                    },
                    options
                );
            }
        };

        this.definitions[name] = {...base, ...overrides};
    }
};












module.exports = {
    requiresVersion: '2.0.0',

    defaults: {},

    showHideTimer: null,

    lockStrings: [],

    _nunjucksEnvironment: null,

    init() {
        Log.log(this.defaults);
    },

    start() {
        Log.info('Starting module: ' + this.name);
    },

    getScripts() {
        return [];
    },

    getStyles() {
        return [];
    },

    getTranslations() {
        return false;
    },

    getDom() {
    },

    getHeader() {
        return this.data.header;
    },

    getTemplate() {
        return '<div class="normal">' + this.name + '</div><div class="small dimmed">' + this.identifier + '</div>';
    },

    getTemplateData() {
        return {};
    },

    notificationReceived(notification, payload, sender) {
        if (sender) {
            Log.log(this.name + ' received a module notification: ' + notification + ' from sender: ' + sender.name);
        } else {
            Log.log(this.name + ' received a system notification: ' + notification);
        }
    },

    nunjucksEnvironment() {
    },

    socketNotificationReceived(notification, payload) {
        Log.log(this.name + ' received a socket notification: ' + notification + ' - Payload: ' + payload);
    },

    suspend() {
        Log.log(this.name + ' is suspended.');
    },

    resume() {
        Log.log(this.name + ' is resumed.');
    },

    setData(data) {
        this.data = data;
        this.name = data.name;
        this.identifier = data.identifier;
        this.hidden = false;

        this.setConfig(data.config, data.configDeepMerge);
    },

    setConfig(config, deep) {
        this.config = deep ? configMerge({}, this.defaults, config) : Object.assign({}, this.defaults, config);
    },

    socket() {
    },

    file(file) {
        return (this.data.path + '/' + file).replace('//', '/');
    },

    loadStyles(callback) {
        this.loadDependencies('getStyles', callback);
    },

    loadScripts(callback) {
        this.loadDependencies('getScripts', callback);
    },

    loadDependencies(funcName, callback) {
    },

    loadTranslations(callback) {
    },

    translate(key, defaultValueOrVariables, defaultValue) {
    },

    updateDom(speed) {
    },

    sendNotification(notification, payload) {
    },

    sendSocketNotification(notification, payload) {
    },

    hide(speed, callback, options) {
        if (typeof callback === 'object') {
            options = callback;
            callback = function () {
            };
        }

        callback = callback || function () {
        };
        options = options || {};

        const self = this;
        MM.hideModule(
            self,
            speed,
            () => {
                self.suspend();
                callback();
            },
            options
        );
    },

    show(speed, callback, options) {
        if (typeof callback === 'object') {
            options = callback;
            callback = function () {
            };
        }

        callback = callback || function () {
        };
        options = options || {};

        const self = this;
        MM.showModule(
            this,
            speed,
            () => {
                self.resume();
                callback();
            },
            options
        );
    }
};
