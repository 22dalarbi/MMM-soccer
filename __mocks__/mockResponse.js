function generateResponse() {
    return {
        season: {
            id: 599,
            currentMatchday: 14,
            startDate: '2020-09-18',
            endDate: '2021-05-15'
        },
        competition: {
            id: 2002,
            area: {id: 2088, name: 'Germany'},
            code: 'BL1',
            name: 'Bundesliga',
            plan: 'TIER_ONE',
            lastUpdated: '2021-01-03T01:25:08Z'
        },
        standings: [
            {
                stage: 'REGULAR_SEASON',
                table: [
                    {
                        position: 1,
                        team: {id: 721, name: 'RB Leipzig', crestUrl: 'https://crests.football-data.org/721.svg'},
                        playedGames: 14,
                        points: 31,
                        won: 9,
                        draw: 4,
                        lost: 1,
                        goalsFor: 25,
                        goalsAgainst: 9,
                        goalDifference: 16,
                        form: 'W,D,W,W,D'
                    },
                    {
                        position: 2,
                        team: {id: 5, name: 'FC Bayern MÃ¼nchen', crestUrl: 'https://crests.football-data.org/5.svg'},
                        playedGames: 13,
                        points: 30,
                        won: 9,
                        draw: 3,
                        lost: 1,
                        goalsFor: 39,
                        goalsAgainst: 19,
                        goalDifference: 20,
                        form: 'W,W,D,D,W'
                    },
                    {
                        position: 3,
                        team: {id: 3, name: 'Bayer 04 Leverkusen', crestUrl: 'https://crests.football-data.org/3.svg'},
                        playedGames: 14,
                        points: 28,
                        won: 8,
                        draw: 4,
                        lost: 2,
                        goalsFor: 29,
                        goalsAgainst: 14,
                        goalDifference: 15,
                        form: 'L,L,W,W,W'
                    }
                ],
                type: 'TOTAL'
            }
        ]
    };
}

module.exports = {generateResponse};
