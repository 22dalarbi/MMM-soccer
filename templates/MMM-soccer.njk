<div class="{% if isModalActive %}MMM-soccer-blur{% endif %}">
    {% if loading %}
        {{'LOADING' | translate}}
    {% else %}
        <header>{{competitionName}}</header>
        <div class="xsmall">{{season}}</div>
        <table class="xsmall table">
            <thead>
                <tr class="row">
                    <th></th>
                    {% if config.logos %}
                        <th></th>
                    {% endif %}
                    <th class="name">{{'TEAM' | translate}}</th>
                    <th class="centered"><i class="fa fa-line-chart"/></th>
                    <th class="centered"><i class="fa fa-soccer-ball-o"/></th>
                </tr>
            </thead>
            <tbody>
                {% for index in range(boundaries.firstTeam, boundaries.lastTeam) %}
                    <tr class="centered-row {% if index === boundaries.focusTeamIndex %}bright{% endif %}" style="{{loop.index | fade(boundaries.focusTeamIndex)}}">
                        <td>{{standing[index].position}}</td>
                        {% if config.logos %}
                            <td><img class="icon {% if not config.colored %}no-color{% endif %}" src="/MMM-soccer/{{standing[index].team.name}}.svg"/></td>
                        {% endif %}
                        <td class="name">{{standing[index].team.name}}</td>
                        <td class="centered">{{standing[index].points}}</td>
                        <td class="centered">{{standing[index].goalDifference}}</td>
                    </tr>
                {% endfor %}
            </tbody>
        </table>
    {% endif %}
</div>
