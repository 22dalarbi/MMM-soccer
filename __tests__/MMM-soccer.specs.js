describe('MMM-soccer', () => {
    beforeAll(() => {
       require('../__mocks__/Logger');
       require('../__mocks__/Module');
    });

    const name = 'MMM-soccer';

    let MMMsoccer;

    beforeEach(() => {
        require('../MMM-soccer');

        MMMsoccer = global.Module.create(name);
        MMMsoccer.setData({name, identifier: `Module_1_${name}`});
    });

    test('requires version 2.14', () => {
        expect(MMMsoccer.requiresVersion).toBe('2.14.0');
    });

    test('has correct defaults', () => {
        expect(MMMsoccer.defaults).toMatchSnapshot();
    });

    test('has correct voice commands', () => {
        expect(MMMsoccer.voice).toMatchSnapshot();
    });

    test('inits module with state loading === true', () => {
        expect(MMMsoccer.loading).toBe(true);
    });

    describe('start', () => {
        let originalInterval = setInterval;

        beforeEach(() => {
            global.setInterval = jest.fn();
        });

        afterEach(() => {
            global.setInterval = originalInterval;
        });

        test('logs start of module', () => {
            MMMsoccer.start();

            expect(global.Log.info).toHaveBeenNthCalledWith(1, 'Starting module: MMM-soccer');
        });

        test('adds nunjuck filters', () => {
            MMMsoccer.start();

            const addFilter = MMMsoccer.nunjucksEnvironment().addFilter;
            expect(addFilter).toBeCalledTimes(1);
            expect(addFilter).toHaveBeenNthCalledWith(1, 'fade', expect.any(Function));
        });

        test('inits current league', () => {
            MMMsoccer.start();

            expect(MMMsoccer.currentLeague).toBe('BL1');
        });

        test('requests data from node_helper', () => {
            MMMsoccer.start();

            expect(MMMsoccer.sendSocketNotification).toHaveBeenNthCalledWith(1, 'GET_DATA', {
                api_key: false, league: 'BL1'
            });
        });

        test('interval requests data from node_helper', () => {
            MMMsoccer.start();

            global.setInterval.mock.calls[0][0]();

            expect(MMMsoccer.sendSocketNotification).toHaveBeenCalledTimes(2);
            expect(MMMsoccer.sendSocketNotification).toHaveBeenNthCalledWith(2, 'GET_DATA', {
                api_key: false, league: 'BL1'
            });
        });

        test('inits interval correctly for usage with api_key', () => {
            MMMsoccer.start();

            expect(global.setInterval).toHaveBeenNthCalledWith(1, expect.any(Function), 1800000);
        });

        test('inits interval correctly for usage without api_key', () => {
            MMMsoccer.setConfig({api_key: 'TEST_API_KEY'});

            MMMsoccer.start();

            expect(global.setInterval).toHaveBeenNthCalledWith(1, expect.any(Function), 300000);
        });
    });
});
