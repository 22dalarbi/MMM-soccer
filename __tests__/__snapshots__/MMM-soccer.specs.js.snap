// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MMM-soccer has correct defaults 1`] = `
Object {
  "api_key": false,
  "colored": false,
  "focus_on": false,
  "leagues": Object {
    "ENGLAND": "PL",
    "FRANCE": "FL1",
    "GERMANY": "BL1",
    "ITALY": "SA",
    "SPAIN": "PD",
  },
  "logos": false,
  "max_teams": false,
  "show": "GERMANY",
}
`;

exports[`MMM-soccer has correct voice commands 1`] = `
Object {
  "mode": "SOCCER",
  "sentences": Array [
    "OPEN HELP",
    "CLOSE HELP",
    "SHOW STANDINGS OF COUNTRY NAME",
    "EXPAND VIEW",
    "COLLAPSE VIEW",
  ],
}
`;
